### ISSUE TITLE
StackOverflowError with erfc

### ISSUE THREAD

**Issue by SebastianM-C** [JuliaIntervals/IntervalArithmetic.jl#113]
Code to reproduce:
```julia
using IntervalArithmetic
using SpecialFunctions
a=0..1
b=0..2
erfc(a/(2*sqrt(b)))
```
This leads to
```julia
ERROR: StackOverflowError:
Stacktrace:
 [1] erfc(::IntervalArithmetic.Interval{Float64}) at /mnt/storage/.julia/v0.6/SpecialFunctions/src/erf.jl:10 (repeats 80000 times)
```

**Comment by lbenet** [comment]
Calling `erfc(a/(2*sqrt(b)))` executes [this line](https://github.com/JuliaMath/SpecialFunctions.jl/blob/032eb893cbae36674f0c16ddf7b17c5089883844/src/erf.jl#L10) in `SpecialFunctions.jl` (since the argument is an interval and Interval{T} <: Real). The problem occurs because `float(a::Interval)` returns an interval, which is used as the argument of `erfc`, again and again... So you get the StackOverflowError.

Since `erfc` is monotonously decreasing, it is not difficult to write a specialized method for intervals. Proper rounding to get a rigorous *tight* interval may be a headache, but the following may be good-enough for your:

```julia
julia> import SpecialFunctions: erfc

julia> erfc(a::Interval{T}) where {T} = interval(prevfloat(erfc(a.hi)), nextfloat(erfc(a.lo)))
erfc (generic function with 12 methods)

julia> erfc(a/(2*sqrt(b)))
[-4.94066e-324, 1.00001]

```

**Comment by lbenet** [comment]
Was this helpful?

**Comment by SebastianM-C** [comment]
Yes. Thank you!
Should I close the issue?

P.s. For compatibility with IntervalConstraintProgramming.jl is there a need for a special treatment for this function?

**Comment by lbenet** [comment]
> P.s. For compatibility with IntervalConstraintProgramming.jl is there a need for a special treatment for this function?

I think there isn't; @dpsanders can you confirm this?

If that solved the issue, you may close it.

**Comment by SebastianM-C** [comment]
I think that it would be helpful to add your answer on defining specialized implementations for intervals for monotonous functions somewhere in the docs. (If this is already documented, could you point me in the right direction?)

For IntervalConstraintProgramming.jl I tried to use a constraint containing `erfc`, but I got `ArgumentError: Function erfc not available. Use @function to define it.`

**Comment by dpsanders** [comment]
IntervalConstraintProgramming requires more information than just the function definition itself, namely an inverse (also called "reverse" or "backward") function. That seems to be available in Julia (`erfcinv`).

So it would be necessary to also add this inverse function to IntervalArithmetic, and then to add a reverse function in `IntervalContractors.jl`. The only problem, as @lbenet hinted, is obtaining *correct* rounding.

**Comment by dpsanders** [comment]
This paper seems to have the implementation:

https://www.sciencedirect.com/science/article/pii/S0890540112000697

Actually I think MPFR may have these functions?

**Comment by dpsanders** [comment]
It only has `erfc`, not the inverse function, so it looks like we would have to implement this.

It may be possible to get a slow one via `ApproxFun` or `TaylorIntegration`.

**Comment by dpsanders** [comment]
Ah, it should be in `ArbFloats`?

**Comment by dpsanders** [comment]
I guess for the inverse function you could use bisection.

**Comment by dpsanders** [comment]
http://fredrikj.net/blog/2016/03/taking-the-error-out-of-the-error-function/

**Comment by SebastianM-C** [comment]
Yes. I would like to set up a list of things that need to be done in order to understand better what to do.

**Comment by dpsanders** [comment]
- In IntervalArithmetic.jl:

  - Implement `erf(x::Interval)` and `erfc(x::Interval)` by using `erf(y::BigFloat)` and `erfc(y::BigFloat)` with monotonicity and directed rounding (as shown above).

  - Implement the inverse functions `erfinv` and `erfcinv` using bisection; must work out how to do directed rounding.

- In IntervalContractors.jl:

  - Follow the pattern that is defined there to add the reverse function `erf_rev`.

**Comment by dpsanders** [comment]
I was searching online and found that you can use the Newton method. Since we know the derivative of the error function, we should be able to use the *interval* Newton method to get a rigorous bound for `erfinv`.

**Comment by SebastianM-C** [comment]
Closed by #118
